sdks/java/extensions/sketching/src/main/java/org/apache/beam/sdk/extensions/sketching/cardinality/ApproximateDistinct.java

```diff
@@ -0,0 +1,352 @@
+      this.p = p;
+      this.sp = sp;
+    }
+
+      /**
+       * Returns an {@code ApproximateDistinctFn} combiner with the given precision value p.
+       * This means that the input elements will be dispatched into 2^p buckets
+       * in order to estimate the cardinality.
+       *
+       * @param p           precision value for the normal representation
+       * @param <InputT>    the type of the Input {@code Pcollection}'s elements being combined.
+       */
+    public static <InputT> ApproximateDistinctFn<InputT> create(int p) {
+      return new ApproximateDistinctFn<>(p, 0);
+    }
+
+      /**
+       * Returns an {@code ApproximateDistinctFn} combiner with the precision value p
+       * of this combiner and the given precision value sp for the sparse representation,
+       * meaning that the combiner will be using a sparse representation at small cardinalities.
+       *
+       * <p>The sparse representation does not initialize every buckets to 0 at the beggining.
+       * Indeed, for small cardinalities a lot of them will remain empty.
+       * Instead it builds a linked-list that grows when new indexes appear in the hashed values.
+       * To reduce collision of indexes and thus improve precision, we can define {@code sp > p}
+       * When the sparse representation would require more memory than the normal one,
+       * it is converted and the normal algorithm applies for the remaining elements.
+       *
+       * <p><b>WARNING : </b>Choose sp such that {@code p <= sp <= 32}
+       *
+       * <p>Example of use :
+       * <pre>{@code PCollection<Integer> input = ...;
+       * PCollection<HyperLogLogPlus> hllSketch = input
+       *        .apply(Combine.globally(ApproximateDistinct.ApproximateDistinctFn.<Integer>create(p)
+       *                .withSparseRepresentation(sp)));
+       * } </pre>
+       *
+       * @param sp          the precision of HyperLogLog+' sparse representation
+       */
+    public ApproximateDistinctFn<InputT> withSparseRepresentation(int sp) {
+      return new ApproximateDistinctFn<>(this.p, sp);
+    }
+
+    @Override
+    public HyperLogLogPlus createAccumulator() {
+      return new HyperLogLogPlus(p, sp);
+    }
+
+    @Override
+    public HyperLogLogPlus addInput(HyperLogLogPlus acc, InputT record) {
+      acc.offer(record);
+      return acc;
+    }
+
+    /**
+     * Output the whole structure so it can be queried, reused or stored easily.
+     */
+    @Override
+    public HyperLogLogPlus extractOutput(HyperLogLogPlus accumulator) {
+      return accumulator;
+    }
+
+    @Override
+    public HyperLogLogPlus mergeAccumulators(Iterable<HyperLogLogPlus> accumulators) {
+      HyperLogLogPlus mergedAccum = createAccumulator();
+      for (HyperLogLogPlus accum : accumulators) {
+        try {
+          mergedAccum.addAll(accum);
+        } catch (CardinalityMergeException e) {
+          // Should never happen because only HyperLogLogPlus accumulators are instantiated.
```

Comment:
<!--new_thread; commit:6bf51fdf6afe6f389415792f335dd08adf9fa85a; resolved:0-->
Throw an IllegalStateException instead of just logging an error otherwise the value that will be returned by mergeAccumulators will be incorrect.